File: Algorithms/445/addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# 方法1：反转链表
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        return reverse(add(reverse(l1), reverse(l2)))


def reverse(head):
    p, prev = head, None
    while p:
        p.next, prev, p = prev, p, p.next
    return prev


def add(l1, l2):
    result = p = ListNode(0)
    carry = 0
    while l1 or l2 or carry:
        x = y = 0
        if l1:
            x, l1 = l1.val, l1.next
        if l2:
            y, l2 = l2.val, l2.next
        carry, val = divmod(x + y + carry, 10)
        p.next = ListNode(val)
        p = p.next
    return result.next

File: 程序员面试金典/2.5/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 迭代版本
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode(0), p = result;
        int carry = 0;
        while (l1 != null || l2 != null) {
            int x = l1 != null ? l1.val : 0;
            int y = l2 != null ? l2.val : 0;
            int s = x + y + carry;
            p.next = new ListNode(s % 10);
            carry = s / 10;
            l1 = l1 != null ? l1.next : l1;
            l2 = l2 != null ? l2.next : l2;
            p = p.next;
        }
        if (carry != 0)
            p.next = new ListNode(carry);
        return result.next;
    }
}
